Support for PyPy3

rahulhacker opened the pull request:
Co-authored-by: SuryanarayanaPeri ss_peri@hotmail.com

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #1033 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [ ] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

ocelotl commented on instrumentation/opentelemetry-instrumentation-system-metrics/src/opentelemetry/instrumentation/system_metrics/__init__.py:
```diff
@@ -108,6 +108,26 @@
 }
 
 
+class MyHooks(object):
```
We are using Python3 only, a class should not inherit from `object`.

rahulhacker replied:
@ocelotl , will make that changes.
Thanks for the review, appreciate.

ocelotl commented on instrumentation/opentelemetry-instrumentation-system-metrics/tests/test_system_metrics.py:
```diff
@@ -18,9 +18,8 @@
-from opentelemetry.instrumentation.system_metrics import (
-    SystemMetricsInstrumentor,
-)
+from opentelemetry.instrumentation.dbapi import SystemMetricsInstrumentor
```
Why this change? All instrumentations tests are failing because of this, by the way.

rahulhacker replied:
@ocelotl : will revert the changes.

ocelotl commented on instrumentation/opentelemetry-instrumentation-system-metrics/src/opentelemetry/instrumentation/system_metrics/__init__.py:
```diff
@@ -108,6 +108,26 @@
+    done = False
+
+    def on_gc_minor(self, stats):
+        print(
```
Why `print`? Are you trying to log something? If so, use a proper logger.

rahulhacker replied:
@ocelotl , it's not required but if anything needs to log will add it accordingly.

ocelotl commented:
Please add a more descriptive name to this PR and also fix lint.

ocelotl commented on instrumentation/opentelemetry-instrumentation-system-metrics/tests/test_system_metrics.py:
```diff
@@ -779,6 +779,22 @@ def test_runtime_get_count(self, mock_gc_get_count):
-        )
+        self._test_metrics(f"runtime.{self.implementation}.gc_count", expected)
+
+    # @mock.patch("gc.hooks")
```
Please don't add code that is commented out.

rahulhacker replied:
@ocelotl : Can you please help us on what all other changes need to be done??

rahulhacker replied:
@ocelotl : Only one of the checks is failing because of below:
ERROR:   pypy3-test-instrumentation-fastapi: commands failed
ERROR:   pypy3-test-instrumentation-starlette: commands failed

Above error is might be because of the Assertion error during the build
>                       self.assertAlmostEqual(duration, point.sum, delta=30)
E                       AssertionError: 35 != 3 within 30 delta (32 difference)

Do need to fix the both one also??
Can you please guide me on this?? It will be really helpful.

rahulhacker commented:
> Please add a more descriptive name to this PR and also fix lint.

Lint issue is fixed

rahulhacker commented:
> We need tests for these changes.

@ocelotl : Added the test cases

rahulhacker commented:
> > We need tests for these changes.
> 
> @ocelotl : Added the test cases

@ocelotl : Some of the checks are failing. Can you please help if missing anything, need some guidance??
As per the understanding we have added the code and test cases accordingly.

SuryanarayanaPeri commented:
@ocelotl - The work in this issue is to make the instrumentation either produce all the same metrics or exclude the instruments that don't have equivalents. Ideally at the end of this issue, the tests for this instrumentation library will be enabled in tox for pypy3.

**For this we have excluded the instruments that don't have equivalents for the gc_count on pypy3. The test for the system metrics instrumentation is enabled in tox. Request you please review and merge the PR.**

We will open a new feature request - to make the instrumentation to provide the gc_count metrics for pypy3.

rahulhacker commented:
> We need tests for these changes.

@ocelotl :Added the test cases to support the pypy3 for system metrics instrumentation.

ocelotl commented on instrumentation/opentelemetry-instrumentation-system-metrics/src/opentelemetry/instrumentation/system_metrics/__init__.py:
```diff
@@ -350,7 +350,7 @@ def _instrument(self, **kwargs):
             )
 
-        if "process.runtime.gc_count" in self._config:
+        if "process.runtime.gc_count" in self._config and self._python_implementation != "pypy":
```
Please add a logger object (like [this](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pymemcache/src/opentelemetry/instrumentation/pymemcache/__init__.py#L53)) and use it to log a warning inside the `__int__` method if the implementation is `pypy`. The warning should let the user know the `process.runtime.gc_count` metric won't be collected because the interpreter is `pypy`.

rahulhacker replied:
@ocelotl : Added the code as per the request. Please have a look and thanks for the prompt response.

ocelotl commented:
Renamed this PR to better reflect the changes being made here.

ocelotl commented:
Ok, almost ready to approve and merge, just take a look at the comment above and implement the requested changes :v:

rahulhacker commented:
@ocelotl : Please don't approve the PR, just hold for one or two days, will let you KNOW and then you can merge the PR

rahulhacker commented:
> @ocelotl : Please don't approve the PR, just hold for one or two days, will let you KNOW and then you can merge the PR, reverting the logger changes
